### PR TITLE
fix: reject path-traversal note ids (arbitrary file read/write/delete)

### PR DESCRIPTION
**Pull request by jamesfishwick**
## Problem

A note id becomes a filename (`{id}.md`), but `Note.id` had no format constraint (only `title` and `references` were validated). Every CRUD MCP tool passes an LLM-supplied id straight into `notes_dir / f"{id}.md"`, so a crafted id escaped the notes directory:

- `id="../../etc/passwd"` climbs out of the notes dir.
- `id="/etc/shadow"` — Python's `Path('/a') / '/etc/x'` **discards the left operand**, so an absolute id ignores `notes_dir` entirely.

Reachable as arbitrary file **read** (`get_note`), **delete** (`delete_note`), and **write** (via the re-index/create path, where a planted file's frontmatter id is trusted and `model_construct` bypasses validation). Note ids come from untrusted content an agent is acting on, so this is not theoretical.

Found by the `/sc:analyze` security pass (Finding 1, CRITICAL).

## Fix — two layers

**Model boundary** (`schema.py`): constrain `Note.id` to `^[A-Za-z0-9_-]+$` (the generated-id alphabet — every real id passes), rejecting `/`, `\`, `.`, and `..` at construction.

**Filesystem boundary** (`note_repository.py`): a `_note_path()` helper resolves the path and asserts `is_relative_to(notes_dir)`, used by `create`/`get`/`update`/`delete`. This refuses a bad id *even if it bypassed model validation* — e.g. the `model_construct` schema-violation hydration path. `_parse_note_from_markdown` additionally skips any file whose frontmatter id is unsafe, closing the `rebuild_index` vector before it can reach `model_construct`.

Defense-in-depth on purpose: either layer alone closes the hole; together they hold even if a future caller constructs a `Note` without validation.

## Tests (`tests/test_note_id_path_traversal.py`, 38 cases)

- Model rejects `../../etc/passwd`, `/etc/shadow`, `..`, `foo/bar`, empty, etc.; accepts generated + manual safe ids.
- Filesystem layer: a **smuggled** (`model_construct`) bad id makes `create`/`update`/`delete`/`get` refuse, with assertions that **no file outside the notes dir** is written, deleted, or read.
- `rebuild_index` skips a planted `.md` whose frontmatter id is `../../tmp/payload`.

Full suite: 379 passed, 1 skipped.

## Scope note

This PR closes only the traversal hole. The `/sc:analyze` pass also flagged multi-process safety (in-process `RLock` + default SQLite engine) and `rebuild_index` atomicity/cost as separate CRITICAL/HIGH items — those are independent and out of scope here.

### CODE REVIEW

File: src/slipbox_mcp/storage/note_repository.py [hunk 1]
```diff
@@ -237,6 +250,15 @@ def _parse_note_from_markdown(self, content: str) -> Optional[Note]:
         note_id = metadata.get("id")
         if not note_id:
             return None
+        if not _is_safe_note_id(note_id):
```
Comment (Korgull-the-Censorious) [inline src/slipbox_mcp/storage/note_repository.py:265]: **[Important] Rebuild-thrash regression introduced by this PR.** This parse-time skip refuses to index an unsafe-id file, but `_count_indexable_files` (~line 205) still *counts* it (its frontmatter matches `^id:\s*\S` and it has no `_is_safe_note_id` gate). So with one unsafe-id `.md` in the notes dir, `indexable_count` includes it while `rebuild_index` drops it, making `db_count != indexable_count` permanently true and firing a FULL rebuild on EVERY `NoteRepository()` construction — every CLI command, every MCP startup. Pre-PR the file indexed cleanly and the counts agreed; this divergence is new. Fix: apply `_is_safe_note_id` in `_count_indexable_files` too, so a file the parser will reject is not counted as indexable.

File: src/slipbox_mcp/storage/note_repository.py [hunk 2]
```diff
@@ -44,6 +44,19 @@ def __init__(self, deleted_id: str, failures: List[tuple]):
         )
 
 
+# Note IDs become filenames (``{id}.md``). Constrain them to a path-safe
+# alphabet so a crafted id (``../../etc/x``, ``/etc/x``) cannot escape the
+# notes dir. This mirrors the validator on Note.id; it is repeated here so the
+# filesystem layer is safe even for ids that bypassed model validation (e.g.
+# Note.model_construct on the schema-violation hydration path).
+_NOTE_ID_PATTERN = re.compile(r"^[A-Za-z0-9_-]+$")
```
Comment (Korgull-the-Censorious) [inline src/slipbox_mcp/storage/note_repository.py]: **[Important] Layer length-divergence — the comment is false.** `_NOTE_ID_PATTERN` is `^[A-Za-z0-9_-]+$` with no length bound, so `_is_safe_note_id('a'*256)` returns True — yet `Note.id` enforces `max_length=255` (schema.py). The header comment claims this pattern mirrors the validator on `Note.id`; that is false re: length. An over-long id passes the filesystem guard but the model would reject it, so the two layers disagree on exactly the thing this PR set out to make consistent. Fix: `^[A-Za-z0-9_-]{1,255}$`, and pin the 255/256 boundary at both layers with a test.

File: src/slipbox_mcp/storage/note_repository.py [hunk 3]
```diff
@@ -237,6 +250,15 @@ def _parse_note_from_markdown(self, content: str) -> Optional[Note]:
         note_id = metadata.get("id")
         if not note_id:
             return None
+        if not _is_safe_note_id(note_id):
+            # A note file whose frontmatter id is not a path-safe stem is
+            # corrupt or hostile (path traversal). Refuse to hydrate it rather
+            # than let it reach the model_construct fallback below, which would
+            # bypass the id validator and make the bad id writable.
+            logger.warning(
+                "Skipping note with unsafe id %r (not a path-safe stem)", note_id
```
Comment (Korgull-the-Censorious) [inline src/slipbox_mcp/storage/note_repository.py]: **[Important] Legacy data loss with a mislabeling log, introduced here.** Ids containing `.` or a space were permitted by the OLD unconstrained schema and may exist in hand-authored notes. On rebuild they now silently vanish from the index with a warning branding them 'corrupt or hostile'. `dot.in.id` is non-conforming, not an attack, yet it's dropped and accused. Blast radius is narrow (generated ids are always safe), so Important not Critical — but distinguish traversal-unsafe (`..`, `/`, `\\`) from merely-non-conforming (`.`, space), and make the log actionable (name the file, point to a remediation) rather than calling the user's own old notes hostile.

File: tests/test_note_id_path_traversal.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+        outside = note_repository.notes_dir.parent / "pwned_sentinel.md"
+        with pytest.raises(ValueError):
+            note_repository.create(_smuggled_note(bad))
+        assert not outside.exists(), "create() wrote a file outside the notes dir"
+
+    def test_update_refuses_unsafe_id(self, note_repository):
+        with pytest.raises(ValueError):
+            note_repository.update(_smuggled_note("../escape"))
+
+    def test_delete_refuses_unsafe_id_and_deletes_nothing(self, note_repository):
```
Comment (Korgull-the-Censorious) [inline tests/test_note_id_path_traversal.py]: **[Important] The worst-case vector is untested at the filesystem boundary.** The docstring emphasizes `Path('/a') / '/etc/x'` discards the left operand — the absolute-path id is the headline threat. Yet this parametrize is only `['../pwned_sentinel', '..', 'foo/bar']`; the absolute id (`/etc/x`, `/tmp/pwned`) is exercised only at the alphabet/model layers, never against create/update/delete/get with a real assertion that no outside file is touched. Add an absolute-id case to the filesystem-boundary tests (assert raise/None AND no outside file written or removed).

File: tests/test_note_id_path_traversal.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+        assert outside.exists(), "delete() removed a file outside the notes dir"
+        outside.unlink()
+
+    def test_get_returns_none_for_unsafe_id(self, note_repository):
+        # Plant a readable file just outside the notes dir.
+        outside = note_repository.notes_dir.parent / "secret.md"
+        outside.write_text("---\nid: secret\ntitle: Secret\n---\ntop secret")
+        assert note_repository.get("../secret") is None
+        outside.unlink()
+
```
Comment (Korgull-the-Censorious) [inline tests/test_note_id_path_traversal.py:167]: **[Important] The highest-stakes write path has the weakest test — and it passes through the wrong branch.** `test_update_refuses_unsafe_id` asserts only that `ValueError` is raised, with no sentinel and no no-side-effect check. Worse: `update()` calls `self.get(note.id)` first, which returns None for an unsafe id and raises 'does not exist' BEFORE `_note_path` is reached. So this test green-lights `update`'s traversal guard while never exercising it. Plant a victim file outside the notes dir and assert its bytes are unchanged; consider making `update`'s guard reachable independent of the `get` pre-check.

File: src/slipbox_mcp/storage/note_repository.py [hunk 6]
```diff
@@ -496,7 +533,9 @@ def create(self, note: Note) -> Note:
 
     def get(self, id: str) -> Optional[Note]:
         """Get a note by ID."""
-        file_path = self.notes_dir / f"{id}.md"
+        if not _is_safe_note_id(id):
```
Comment (Korgull-the-Censorious) [inline src/slipbox_mcp/storage/note_repository.py:558]: **[Nit] Redundant guard worth a comment.** `get()` pre-checks `_is_safe_note_id` and returns None, then calls `_note_path` which re-checks and raises. This is a deliberate raise-vs-None seam (None for a reader, raise as internal last-resort) and it's correct — but it looks like duplication a future 'simplification' would collapse, removing a layer. Add a one-line comment that the double-check is intentional defense-in-depth.

File: tests/test_note_id_path_traversal.py [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+class TestRebuildIndexVector:
+    def test_planted_traversal_id_file_is_skipped(self, note_repository):
+        """A planted .md whose frontmatter id is a traversal stub must not index.
+
+        This is the rebuild_index attack path: the parser refuses the hostile
+        id before it can reach the model_construct fallback (which would bypass
+        the validator and make the bad id writable on a later update()).
+        """
+        planted = note_repository.notes_dir / "seed.md"
+        planted.write_text(
```
Comment (Korgull-the-Censorious) [inline tests/test_note_id_path_traversal.py:207]: **[Nit] The symlink-escape vector is defended but untested.** The `is_relative_to` check is load-bearing for a symlink INSIDE notes_dir whose name is a valid stem but resolves outside — the regex passes, only `.resolve()` + `is_relative_to` catches it. No test exercises this, so if that line were dropped in a refactor the suite stays green while the guarantee is gone. Plant a valid-stem symlink pointing outside notes_dir and assert get/create/delete refuse it.

**Review by Korgull-the-Censorious — CHANGES_REQUESTED** [review]
# The Re-Extermination — PR #29 (path-traversal guard)

I walked your guard across the wasted plain with four blind lenses. The traversal defense itself HOLDS — two lenses tried to break the alphabet + `.resolve()`/`is_relative_to` and could not, on read and write alike. But the change bleeds collateral, and that is where I found the survivors.

## Verdict
**Request changes.** Two Important regressions ride in with the fix: a rebuild-thrash where `_count_indexable_files` counts unsafe-id files the new parse-skip refuses to index (full rebuild on every construction), and a layer divergence where the repo regex has no length cap while the schema enforces `max_length=255` — so the comment claiming it 'mirrors the validator' lies about length. Plus two Important test/data-loss gaps and two nits. Tally: 5 Important, 2 Nit. The defense is sound; its blast radius is not. Findings live inline on the lines that bear them.

Korgull rules the earth.

File: src/slipbox_mcp/storage/note_repository.py (hunk 1 above)
Comment (jamesfishwick) [reply src/slipbox_mcp/storage/note_repository.py:265]: Fixed in 816f244. `_count_indexable_files` now extracts the id and gates it with `_is_safe_note_id`, so a file the parser will skip is not counted as indexable. Added `test_unsafe_id_file_not_counted_as_indexable` pinning count==1 with a planted unsafe-id file present.

File: src/slipbox_mcp/storage/note_repository.py (hunk 2 above)
Comment (jamesfishwick) [reply src/slipbox_mcp/storage/note_repository.py]: Fixed in 816f244. `_NOTE_ID_PATTERN` is now `^[A-Za-z0-9_-]{1,255}$` and the comment updated to claim the parity it now actually has. Added length tests at both layers (255 accepted, 256 rejected).

File: src/slipbox_mcp/storage/note_repository.py (hunk 3 above)
Comment (jamesfishwick) [reply src/slipbox_mcp/storage/note_repository.py]: Partially accepted in 816f244. Kept the strict alphabet (intentional hygiene — generated ids are always [0-9T] and never trip it), but reworded the warning: it names the note, no longer brands legacy ids 'hostile', and points at the rename remediation. Widening the alphabet would reintroduce the exact model/repo divergence flagged in the length comment, so I made the skip loud and actionable rather than silent.

File: tests/test_note_id_path_traversal.py (hunk 4 above)
Comment (jamesfishwick) [reply tests/test_note_id_path_traversal.py]: Added in 816f244. New `test_create_refuses_absolute_id_and_writes_nothing`, `test_update_refuses_absolute_id`, absolute variants of delete/get, and `test_planted_absolute_id_file_is_skipped` — each asserts no file outside the notes dir is written, removed, or read.

File: tests/test_note_id_path_traversal.py (hunk 5 above)
Comment (jamesfishwick) [reply tests/test_note_id_path_traversal.py:167]: Fixed in 816f244. `update()` now calls `_note_path(note.id)` before `get()`, so the containment guard is reachable instead of masked by 'does not exist'. Test renamed to `test_update_refuses_unsafe_id_and_overwrites_nothing`: plants a victim outside and asserts its bytes are unchanged.

File: src/slipbox_mcp/storage/note_repository.py (hunk 6 above)
Comment (jamesfishwick) [reply src/slipbox_mcp/storage/note_repository.py:558]: Done in 816f244. Added a comment marking the double-check intentional (None for a reader, raise as the write-path last-resort) so a future simplification doesn't collapse one layer.

File: tests/test_note_id_path_traversal.py (hunk 7 above)
Comment (jamesfishwick) [reply tests/test_note_id_path_traversal.py:207]: Added in 816f244. `test_symlink_with_valid_stem_resolving_outside_is_refused` plants a symlink `evil.md` (valid stem) pointing outside notes_dir and asserts `get('evil')` raises — pinning the `.resolve()` + `is_relative_to` guard.

File: src/slipbox_mcp/storage/note_repository.py [hunk 8]
```diff
@@ -189,7 +203,12 @@ def _count_indexable_files(self) -> int:
                 if end == -1:
                     continue
                 frontmatter_block = header[3:end]
-                if re.search(r"^id:\s*\S", frontmatter_block, re.MULTILINE):
+                m = re.search(r"^id:\s*(\S+)", frontmatter_block, re.MULTILINE)
```
Comment (Korgull-the-Censorious) [inline src/slipbox_mcp/storage/note_repository.py]: **[Important — STILL BREATHING] The fix approximates the parser instead of matching it.** `_count_indexable_files` extracts the id with this token regex `^id:\s*(\S+)` + quote-strip, but `_parse_note_from_markdown` uses `frontmatter.loads` (YAML). They diverge on realistic ids, each re-creating the rebuild-thrash: `id: 12345` → YAML yields an int, `_is_safe_note_id(12345)` is False so the parser SKIPS, but this regex stringifies `12345` and COUNTS it; `id: "foo bar"` → parser yields `foo bar` (space → skipped) while `(\S+)` grabs `"foo`, strips the quote to `foo` and counts it. Either makes `db_count != indexable_count` permanently true → full rebuild on every construction. The off-by-one the original finding warned about (quoting, whitespace, extraction) is exactly what survived. Fix: parse the frontmatter block the SAME way the parser does (`frontmatter.loads`/`yaml.safe_load` on the header) and gate on `_is_safe_note_id(parsed_id)` — don't approximate the YAML parse with a token regex. Add a regression test with `id: 12345` and `id: "foo bar"`.

File: tests/test_note_id_path_traversal.py [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+
+    def test_update_refuses_absolute_id(self, note_repository):
+        target = note_repository.notes_dir.parent / "abs_update_payload"
+        with pytest.raises(ValueError):
+            note_repository.update(_smuggled_note(str(target)))
+        assert not target.parent.joinpath("abs_update_payload.md").exists()
+
+    @pytest.mark.parametrize("bad", ["../victim", None])  # None -> absolute, set below
+    def test_delete_refuses_unsafe_id_and_deletes_nothing(self, note_repository, bad):
+        outside = note_repository.notes_dir.parent / "victim.md"
```
Comment (Korgull-the-Censorious) [inline tests/test_note_id_path_traversal.py]: **[Nit] FS-boundary tests plant into the shared temp root.** `notes_dir.parent` is the shared system temp dir (TemporaryDirectory's parent), and these plant fixed names (`victim.md`, `secret.md`, `pwned_sentinel.md`). Two concurrent test runs — or a leftover from a crashed run — can collide on those names and flake or false-pass. Prefer a dedicated second TemporaryDirectory, or derive the planted name from the unique `notes_dir.name` plus a matching traversal id.

File: src/slipbox_mcp/storage/note_repository.py [hunk 10]
```diff
@@ -44,6 +44,20 @@ def __init__(self, deleted_id: str, failures: List[tuple]):
         )
 
 
+# Note IDs become filenames (``{id}.md``). Constrain them to a path-safe
+# alphabet so a crafted id (``../../etc/x``, ``/etc/x``) cannot escape the
+# notes dir. This mirrors the validator on Note.id exactly -- same alphabet AND
+# the same 1..255 length bound -- and is repeated here so the filesystem layer
+# is safe even for ids that bypassed model validation (e.g. Note.model_construct
+# on the schema-violation hydration path).
+_NOTE_ID_PATTERN = re.compile(r"^[A-Za-z0-9_-]{1,255}$")
```
Comment (Korgull-the-Censorious) [inline src/slipbox_mcp/storage/note_repository.py:53]: **[Resolved]** Length parity bites: `_NOTE_ID_PATTERN = ^[A-Za-z0-9_-]{1,255}$`, `_is_safe_note_id('a'*256)` now False / `'a'*255` True, and the comment truthfully claims the 1..255 bound. Locked by `test_length_bound_matches_model` + `test_model_rejects_overlong_id`.

File: src/slipbox_mcp/storage/note_repository.py [hunk 11]
```diff
@@ -237,6 +256,21 @@ def _parse_note_from_markdown(self, content: str) -> Optional[Note]:
...
+        if not _is_safe_note_id(note_id):
+            # The id is not an accepted stem: either a path-traversal attempt
+            # (``../../etc/x``, ``/etc/x``) or merely a legacy id using chars
+            # outside [A-Za-z0-9_-] (e.g. a dot or space from before the id was
+            # constrained). Skip it rather than let it reach the model_construct
+            # fallback below, which would bypass the id validator and make a
+            # traversal id writable. Skipping is loud and actionable so a benign
+            # legacy note is not lost silently -- rename its id to re-index it.
+            logger.warning(
+                "Skipping note %s: id %r is not an accepted stem "
```
Comment (Korgull-the-Censorious) [inline src/slipbox_mcp/storage/note_repository.py:274]: **[Resolved]** The warning no longer slanders legacy notes: it names the note, states the allowed stem, and gives the rename remediation instead of branding benign ids 'hostile'. Strict alphabet kept by design — defensible as the traversal-safety boundary; the silent/mislabeled failure mode is gone.

File: tests/test_note_id_path_traversal.py [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+
+class TestFilesystemBoundary:
+    @pytest.mark.parametrize("bad", ["../pwned_sentinel", "..", "foo/bar"])
+    def test_create_refuses_unsafe_id_and_writes_nothing(self, note_repository, bad):
+        outside = note_repository.notes_dir.parent / "pwned_sentinel.md"
+        with pytest.raises(ValueError):
+            note_repository.create(_smuggled_note(bad))
+        assert not outside.exists(), "create() wrote a file outside the notes dir"
+
+    def test_create_refuses_absolute_id_and_writes_nothing(self, note_repository):
```
Comment (Korgull-the-Censorious) [inline tests/test_note_id_path_traversal.py:121]: **[Resolved]** The operand-discard worst case is now pinned: absolute-id tests for create/update/delete/get, each asserting no outside file is written/removed/read — not merely that ValueError is raised. Plus `test_planted_absolute_id_file_is_skipped`.

File: src/slipbox_mcp/storage/note_repository.py [hunk 13]
```diff
@@ -528,13 +583,17 @@ def get_all(self) -> List[Note]:
 
     def update(self, note: Note) -> Note:
         """Update a note."""
+        # Resolve+contain the path first, so a smuggled traversal id (one that
+        # bypassed model validation via model_construct) is refused loudly here
+        # rather than being masked by the "does not exist" check below.
+        file_path = self._note_path(note.id)
```
Comment (Korgull-the-Censorious) [inline src/slipbox_mcp/storage/note_repository.py:595]: **[Resolved]** `_note_path(note.id)` is now the first statement of `update()`, before `get()`, so the containment guard is reachable rather than masked by 'does not exist'. The test plants a victim outside and asserts its bytes are unchanged.

File: src/slipbox_mcp/storage/note_repository.py [hunk 14]
```diff
@@ -496,7 +545,13 @@ def create(self, note: Note) -> Note:
 
     def get(self, id: str) -> Optional[Note]:
         """Get a note by ID."""
-        file_path = self.notes_dir / f"{id}.md"
+        # Intentional double-check (do not collapse): a reader asking for a bad
```
Comment (Korgull-the-Censorious) [inline src/slipbox_mcp/storage/note_repository.py:554]: **[Resolved]** The intentional double-check is documented (None for a reader, raise as the write-path last-resort), so a future simplification won't collapse a layer.

File: tests/test_note_id_path_traversal.py [hunk 15]
```diff
@@ -0,0 +1,229 @@
...
+
+    def test_get_returns_none_for_unsafe_id(self, note_repository):
+        # Plant a readable file just outside the notes dir.
+        outside = note_repository.notes_dir.parent / "secret.md"
+        outside.write_text("---\nid: secret\ntitle: Secret\n---\ntop secret")
+        assert note_repository.get("../secret") is None
+        assert note_repository.get(str(note_repository.notes_dir.parent / "secret")) is None
+        outside.unlink()
+
+    def test_symlink_with_valid_stem_resolving_outside_is_refused(self, note_repository):
```
Comment (Korgull-the-Censorious) [inline tests/test_note_id_path_traversal.py:168]: **[Resolved]** The symlink-escape vector is pinned: a valid-stem symlink pointing outside notes_dir, asserting `get('evil')` raises — genuinely exercising the `.resolve()`+`is_relative_to` branch the regex can't catch.

**Review by Korgull-the-Censorious — CHANGES_REQUESTED** [review]
# The Re-Extermination, Pass 2 — PR #29

A fresh skull, blind to the author's reasoning, walked the fixes. Six of seven heads are off clean: the length bound bites, the warning stopped slandering legacy notes, update()'s guard is reachable and side-effect-tested, the absolute-id and symlink vectors are pinned with real no-side-effect assertions, the double-check is documented.

## Verdict
**Request changes.** One survivor: the rebuild-thrash fix approximates the parser with a token regex instead of matching its `frontmatter.loads`, so `id: 12345` and `id: "foo bar"` still count-but-don't-index and thrash a full rebuild on every construction — reproduced live. The off-by-one the original finding warned about survived the patch. 1 Important + 1 Nit remain; 6 Resolved.

Korgull rules the earth.

File: src/slipbox_mcp/storage/note_repository.py (hunk 8 above)
Comment (jamesfishwick) [reply src/slipbox_mcp/storage/note_repository.py]: Fixed in 2691e3a. `_count_indexable_files` now parses the frontmatter block with `frontmatter.loads(...)` and gates on `_is_safe_note_id(meta.get('id'))` — the exact mechanism `_parse_note_from_markdown` uses — so the count and parser agree on YAML-typed and quoted ids. Added `test_count_matches_parser_for_tricky_ids` parametrized over `12345`, `\"foo bar\"`, `'a b'`, `../x`, asserting `_count_indexable_files() == db_count` after a real rebuild. Good catch — the token regex was a false equivalence.

File: tests/test_note_id_path_traversal.py (hunk 9 above)
Comment (jamesfishwick) [reply tests/test_note_id_path_traversal.py]: Fixed in 2691e3a. The filesystem-boundary tests now derive every planted outside name from the unique per-test `notes_dir.name` (with matching traversal ids), so concurrent runs and crashed-run leftovers can't collide on a shared-temp-root filename.

File: src/slipbox_mcp/storage/note_repository.py [hunk 16]
```diff
@@ -188,8 +202,19 @@ def _count_indexable_files(self) -> int:
...
+                # Extract the id EXACTLY as _parse_note_from_markdown does --
+                # via frontmatter (YAML), not a token regex -- so the count and
+                # the parser agree on what is indexable. A token regex diverges
+                # on YAML-typed ids (`id: 12345` -> int, parser skips) and
+                # quoted ids with spaces (`id: "foo bar"`), and any divergence
+                # makes db_count != indexable_count permanently true, thrashing a
+                # full rebuild on every construction. Count only what the parser
+                # will actually index: a present, path-safe id.
+                try:
+                    meta = frontmatter.loads(header[:end] + "\n---\n").metadata
```
Comment (Korgull-the-Censorious) [inline src/slipbox_mcp/storage/note_repository.py:214]: **[Resolved] Finding 1 — rebuild-thrash is dead.** `_count_indexable_files` now reads the id through the same eye as the parser: `frontmatter.loads(header[:end] + '\n---\n').metadata` gated by `_is_safe_note_id`. I walked `id: 12345` (int -> both skip), `"foo bar"`/`'a b'` (space -> both skip), a normal timestamp id (both index), and a missing id (both skip) — count and parser never disagree now, so `db_count == indexable_count` holds and the every-construction rebuild is gone. Locked by `test_count_matches_parser_for_tricky_ids`.

File: tests/test_note_id_path_traversal.py [hunk 17]
```diff
@@ -0,0 +1,263 @@
...
+    now = datetime.datetime.now()
+    return Note.model_construct(
+        id=bad_id, title="t", content="c", note_type=NoteType.PERMANENT,
+        tags=[], links=[], references=[], created_at=now, updated_at=now,
+        metadata={},
+    )
+
+
+class TestFilesystemBoundary:
+    # Planted "outside" files use a name derived from the unique per-test
```
Comment (Korgull-the-Censorious) [inline tests/test_note_id_path_traversal.py:104]: **[Resolved] Nit — fixtures no longer claw at a shared corpse.** Every planted 'outside' file now derives its name from the unique per-run `notes_dir.name` with a matching traversal id, so concurrent runs and crashed-run leftovers can't collide in the shared temp root.

**Review by Korgull-the-Censorious — APPROVED** [review]
# The Re-Extermination, Pass 3 — PR #29

A third skull walked it blind. The last survivor — the count reading ids through a different eye than the parser — is dead: I tried int ids, quoted-space ids, nulls and floats, and the two paths condemned the same files every time. The fixtures now brand each victim with the run's unique temp stem.

## Verdict
**Approved.** No Critical or Important wound remains. The traversal defense was sound from the first pass; the collateral it bled — rebuild-thrash, layer divergence, mislabeled data loss, untested worst-cases — has been cauterized over three passes. The plain is quiet. Merge it.

Korgull rules the earth.
